src/main/java/betteradvancements/util/CriterionGrid.java

```diff
@@ -0,0 +1,132 @@
+            int columnWidth = 0;
+            for (int r = 0; r < numRows; r++) {
+                int cellIndex = c * numRows + r;
+                if (cellIndex >= cellContents.size()) {
+                    break;
+                }
+                String str = cellContents.get(cellIndex);
+                column.add(str);
+                columnWidth = Math.max(columnWidth, cellWidths[cellIndex]);
+            }
+            columns.add(new Column(column, columnWidth));
+            widthSum += columnWidth;
+        }
+        this.columns = columns;
+        this.width = widthSum;
+        this.height = numRows * fontHeight;
+        this.aspectRatio = width / height;
+    }
+
+    public class Column {
+        public final List<String> cells;
+        public final int width;
+
+        public Column(List<String> cells, int width) {
+            this.cells = cells;
+            this.width = width;
+        }
+    }
+
+    public static CriterionGrid findOptimalCriterionGrid(Advancement advancement, AdvancementProgress progress, double maxAspectRatio, FontRenderer renderer) {
+        if (progress == null || progress.isDone()) {
+            return new CriterionGrid();
+        }
+        Map<String, Criterion> criteria = advancement.getCriteria();
+        if (criteria.size() <= 1) {
+            return new CriterionGrid();
+        }
+        boolean anyObtained = false;
+        int numUnobtained = 0;
+        List<String> cellContents = new ArrayList<String>();
+        for (String criterion : criteria.keySet()) {
+            if (progress.getCriterionProgress(criterion).isObtained()) {
+                cellContents.add(" �2+�  " + criterion);
+                anyObtained = true;
+            }
+            else {
+                if (showSpoilers) {
+                    cellContents.add(" �4x�  " + criterion);
+                }
+                numUnobtained++;
+            }
+        }
+        if (!anyObtained) {
+            return new CriterionGrid();
+        }
+        if (!showSpoilers) {
+            cellContents.add(" �4x�  �o" + numUnobtained + " remaining");
+        }
+
+        int[] cellWidths = new int[cellContents.size()];
+        for (int i = 0; i < cellWidths.length; i++) {
+            cellWidths[i] = renderer.getStringWidth(cellContents.get(i));
+        }
+
+        CriterionGrid prevGrid = null;
+        for (int numCols = 1; numCols <= cellContents.size(); numCols++)
+        {
+            CriterionGrid currGrid = new CriterionGrid(cellContents, cellWidths, renderer.FONT_HEIGHT, numCols);
+            if (prevGrid != null && currGrid.numRows == prevGrid.numRows)
+                continue; // We increased the width without decreasing the height, which is pointless.
+            currGrid.init();
+            if (currGrid.aspectRatio > maxAspectRatio) {
+                if (prevGrid == null) {
+                    prevGrid = currGrid;
+                }
+                break;

```

Comment:
Considering you are breaking here, wouldn't it be better to use a `while`? Not that I have anything against breaking, but it does make the code harder to read. If making this a `while` makes it more complex than leave it as is.

Reply:
I think a do-while worked out :)